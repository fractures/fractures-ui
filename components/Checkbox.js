import cc from "classcat"
import Label from "./shared/Label"
import React, { Component } from "react"

const defaultProps = {
	check: () => null,
	isChecked: false,
	isDisabled: false,
	label: ""
}

class Checkbox extends Component {
	render() {
		const checkboxClasses = cc({
			"fr-checkbox flex flex-gap-1 relative": true,
			"fr-checkbox--checked": this.props.isChecked,
			"fr-checkbox--unchecked": !this.props.isChecked,
			"fr-checkbox--disabled": this.props.isDisabled
		})

		return (
			<label className={ checkboxClasses }>
				<input
					type="checkbox"
					defaultChecked={ this.props.isChecked }
					onChange={ () => this.props.check() }
					className="h-0 w-0 none absolute"
				/>
				<span className="fr-checkbox__mark" />
				<Label>{this.props.label}</Label>
				<style jsx>{`
					.fr-checkbox {
						cursor: pointer;
						user-select: none;
					}

					.fr-checkbox__mark {
						height: 1.75rem;
						overflow: hidden;
						width: 1.75rem;

						background-color: var(--shade-0);
						border-radius: 0.125rem;
						border: 2px solid var(--shade-5);
						transition: border-color 120ms ease-in-out;
					}

					.fr-checkbox__mark:after {
						left: 0.625rem;
						top: 0.25rem;
						position: absolute;

						height: 1rem;
						opacity: 0;
						width: 0.5rem;

						border: solid var(--shade-5);
						border-width: 0 0.1875rem 0.1875rem 0;
						transform: rotate(45deg) translateY(-0.125rem) translateX(-0.125rem);
						transition: all 120ms ease-in-out;

						content: " ";
					}

					.fr-checkbox input:checked ~ .fr-checkbox__mark:after {
						opacity: 1;

						transform: rotate(45deg) translateY(0) translateX(0);
					}

					.fr-checkbox input:checked ~ .fr-checkbox__mark {
						background-color: var(--shade-0);
					}

					.fr-checkbox--disabled {
						opacity: 0.5;
						pointer-events: none;
					}
				`}</style>
			</label>
		)
	}
}

Checkbox.defaultProps = defaultProps

export default Checkbox
