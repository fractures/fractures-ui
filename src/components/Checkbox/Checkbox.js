import cc from "classcat"
import Label from "../Label/Label"
import PropTypes from "prop-types"
import React from "react"
import styled from "styled-components"

const propTypes = {
	check: PropTypes.func.isRequired,
	className: PropTypes.string,
	isChecked: PropTypes.bool,
	isDisabled: PropTypes.bool,
	label: PropTypes.string.isRequired
}

const defaultProps = {
	check: () => null,
	isChecked: false,
	isDisabled: false,
	label: ""
}

const NakedCheckbox = props => {
	const checkboxClasses = cc({
		"fr-checkbox flex flex-gap-1 relative": true,
		"fr-checkbox--checked": props.isChecked,
		"fr-checkbox--unchecked": !props.isChecked,
		"fr-checkbox--disabled": props.isDisabled
	})

	return (
		<div className={ props.className }>
			<label className={ checkboxClasses }>
				<input type="checkbox" defaultChecked={ props.isChecked } onChange={ () => props.check() } className="h-0 w-0 none" />
				<span className="fr-checkbox__mark" />
				<Label>{props.label}</Label>
			</label>
		</div>
	)
}

const Checkbox = styled(NakedCheckbox)`
	cursor: pointer;
	user-select: none;

	& .fr-checkbox--disabled {
		opacity: 0.5;
		pointer-events: none;
	}

	& .fr-checkbox__mark {
		height: 1.75rem;
		overflow: hidden;
		width: 1.75rem;

		background-color: var(--fr-ground);
		border-radius: 0.125rem;
		border: 2px solid var(--fr-500);
		transition: border-color 120ms ease-in-out;
	}

	& .fr-checkbox__mark:after {
		left: 0.625rem;
		top: 0.25rem;
		position: absolute;

		height: 1rem;
		opacity: 0;
		width: 0.5rem;

		border: solid var(--fr-500);
		border-width: 0 0.1875rem 0.1875rem 0;
		transform: rotate(45deg) translateY(-0.125rem) translateX(-0.125rem);
		transition: all 120ms ease-in-out;
		content: " ";
	}

	& input:checked ~ .fr-checkbox__mark:after {
		opacity: 1;
		transform: rotate(45deg) translateY(0) translateX(0);
	}

	& input:checked ~ .fr-checkbox__mark {
		background-color: var(--fr-ground);
	}
`

const Toggle = styled(NakedCheckbox)`
	cursor: pointer;
	user-select: none;

	& .fr-checkbox--disabled {
		opacity: 0.5;
		pointer-events: none;
	}

	& .fr-checkbox__mark {
		height: 1.5rem;
		overflow: hidden;
		width: 3rem;

		background-color: var(--fr-500);
		border-radius: 3rem;
	}

	& .fr-checkbox__mark:after {
		left: 0.25rem;
		top: 0.25rem;
		position: absolute;

		height: 1rem;
		width: 1rem;

		background-color: white;
		border-radius: 100%;
		transition: all 120ms ease-in-out;
		content: " ";
	}

	& .fr-checkbox__mark:before {
		left: 0;
		top: 0;
		position: absolute;

		padding: 0 0.5rem;
		width: 3rem;

		color: white;

		content: "Off";
		font-size: 0.75rem;
		font-weight: 500;
		line-height: 1.5rem;
		text-align: right;
	}

	& input:checked ~ .fr-checkbox__mark:after {
		left: 1.75rem;
	}

	& input:checked ~ .fr-checkbox__mark:before {
		content: "On";
		text-align: left;
	}

	& input:checked ~ .fr-checkbox__mark {
		background-color: var(--fr-700);
	}
`

NakedCheckbox.propTypes = propTypes
NakedCheckbox.defaultProps = defaultProps

Checkbox.propTypes = propTypes
Checkbox.defaultProps = defaultProps

Toggle.propTypes = propTypes
Toggle.defaultProps = defaultProps

export { Checkbox, NakedCheckbox, Toggle }
