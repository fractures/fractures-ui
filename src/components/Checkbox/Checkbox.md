```js noeditor
<Fractures theme="blue" />
```

#### Checkbox non-checked

```js
<Checkbox label="Checkbox" isChecked={ false } />
```

#### Checkbox non-checked and disabled

```js
<Checkbox label="Checkbox" isChecked={ false } isDisabled={ true } />
```

#### Checkbox checked

```js
<Checkbox label="Checkbox" isChecked={ true } />
```

#### Checkbox checked and disabled

```js
<Checkbox label="Checkbox" isChecked={ true } isDisabled={ true } />
```
