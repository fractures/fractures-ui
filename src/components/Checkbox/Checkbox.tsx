import classNames from '@sindresorhus/class-names'
import Label from '../shared/Label/Label'
import PropTypes from 'prop-types'
import React from 'react'
import styled from 'styled-components'

const defaultProps = {
	onChange: () => null,
	isChecked: false,
	isDisabled: false,
	label: ''
}

const NakedCheckbox = props => {
	const checkboxClasses = classNames('hl-checkbox', {
		'hl-checkbox--checked': props.isChecked,
		'hl-checkbox--unchecked': !props.isChecked,
		'hl-checkbox--disabled': props.isDisabled
	})

	return (
		<div className={ props.className }>
			<label className={ checkboxClasses }>
				<input type="checkbox" defaultChecked={ props.isChecked } onChange={ () => props.onChange() } />
				<span className="hl-checkbox__mark" />
				<Label>{props.label}</Label>
			</label>
		</div>
	)
}

NakedCheckbox.defaultProps = defaultProps

const Checkbox = styled(NakedCheckbox)`
	box-sizing: border-box;

	user-select: none;

	label {
		cursor: pointer;
	}

	input {
		display: none;
		height: 0;
		width: 0;

		visibility: none;
	}

	.hl-checkbox {
		position: relative;

		display: flex;
	}

	.hl-checkbox__mark {
		display: flex;
		height: 1.5rem;
		margin-right: 0.5rem;
		overflow: hidden;
		min-width: 1.5rem;
		width: 1.5rem;

		background-color: var(--color-gray-0);
		border-radius: 0.125rem;
		border: 2px solid var(--color-gray-500);
		transition: border-color 100ms ease-in-out;
	}

	.hl-checkbox__mark:after {
		left: 0.625rem;
		top: 0.25rem;
		position: absolute;

		box-sizing: inherit;
		height: 0.875rem;
		opacity: 0;
		width: 0.375rem;

		border: solid var(--color-gray-500);
		border-width: 0 0.1875rem 0.1875rem 0;
		transform: rotate(45deg) translateY(-0.125rem) translateX(-0.125rem);
		transition: all 100ms ease-in-out;
		content: ' ';
	}

	.hl-checkbox--disabled {
		opacity: 0.5;
		pointer-events: none;
	}

	input:checked ~ .hl-checkbox__mark:after {
		opacity: 1;
		transform: rotate(45deg) translateY(0) translateX(0);
	}

	input:checked ~ .hl-checkbox__mark {
		background-color: var(--color-gray-0);
	}
`



export default Checkbox
