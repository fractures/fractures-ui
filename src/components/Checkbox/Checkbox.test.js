import Checkbox from "./Checkbox.js"

const mockFn = jest.fn()

describe("Checkbox", () => {
	const componentRender = <Checkbox label="Checkbox" check={ mockFn() } />
	const component = renderer.create(componentRender)
	const tree = component.toJSON()

	it("Renders correctly", () => {
		expect(tree).toMatchSnapshot()
	})
})

describe("Checkbox checked", () => {
	const componentRender = <Checkbox label="Checked checkbox" isChecked={ true } check={ mockFn() } />
	const component = renderer.create(componentRender)
	const tree = component.toJSON()

	it("Renders correctly", () => {
		expect(tree).toMatchSnapshot()
	})
})

describe("Checkbox disabled", () => {
	const componentRender = <Checkbox label="Disabled checkbox" isDisabled={ true } check={ mockFn() } />
	const component = renderer.create(componentRender)
	const tree = component.toJSON()

	it("Renders correctly", () => {
		expect(tree).toMatchSnapshot()
	})
})
