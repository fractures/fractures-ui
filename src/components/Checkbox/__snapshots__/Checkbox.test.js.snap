// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox Renders correctly 1`] = `
<div
  className="sc-bwzfXH diQmhH"
>
  <label
    className="fr-checkbox fr-checkbox--unchecked"
  >
    <input
      defaultChecked={false}
      onChange={[Function]}
      type="checkbox"
    />
    <span
      className="fr-checkbox__mark"
    />
    <span
      className="sc-bdVaJa erEhky"
    >
      Checkbox
    </span>
  </label>
</div>
`;

exports[`Checkbox checked Renders correctly 1`] = `
<div
  className="sc-bwzfXH diQmhH"
>
  <label
    className="fr-checkbox fr-checkbox--checked"
  >
    <input
      defaultChecked={true}
      onChange={[Function]}
      type="checkbox"
    />
    <span
      className="fr-checkbox__mark"
    />
    <span
      className="sc-bdVaJa erEhky"
    >
      Checked checkbox
    </span>
  </label>
</div>
`;

exports[`Checkbox disabled Renders correctly 1`] = `
<div
  className="sc-bwzfXH diQmhH"
>
  <label
    className="fr-checkbox fr-checkbox--unchecked fr-checkbox--disabled"
  >
    <input
      defaultChecked={false}
      onChange={[Function]}
      type="checkbox"
    />
    <span
      className="fr-checkbox__mark"
    />
    <span
      className="sc-bdVaJa erEhky"
    >
      Disabled checkbox
    </span>
  </label>
</div>
`;
